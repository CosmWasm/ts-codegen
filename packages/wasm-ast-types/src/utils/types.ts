import * as t from '@babel/types';
import { camel, pascal, snake } from 'case';
import { propertySignature } from './babel';
import { RenderContext } from '../context';
import { JSONSchema } from '../types';

export function getResponseType(
  context: RenderContext,
  underscoreName: string
) {
  const methodName = camel(underscoreName);
  return pascal(
    context.contract?.responses?.[underscoreName]?.title
    ??
    // after v1.1 is adopted, we can deprecate this and require the above response
    `${methodName}Response`
  );
};

const getTypeStrFromRef = ($ref) => {
  if ($ref?.startsWith('#/definitions/')) {
    return $ref.replace('#/definitions/', '');
  }
  throw new Error('what is $ref: ' + $ref);
}

export const getTypeFromRef = ($ref) => {
  return t.tsTypeReference(t.identifier(getTypeStrFromRef($ref)))
}

const getArrayTypeFromRef = ($ref) => {
  return t.tsArrayType(
    getTypeFromRef($ref)
  );
}

const getTypeOrRef = (obj) => {
  if (obj.type) {
    return getType(obj.type)
  }
  if (obj.$ref) {
    return getTypeFromRef(obj.$ref);
  }
  throw new Error('contact maintainers cannot find type for ' + obj)
}

const getArrayTypeFromItems = (items) => {
  // passing in [{"type":"string"}]
  if (Array.isArray(items)) {
    if (items[0]?.type === 'array') {
      return getArrayTypeFromItems(
        items[0]
      );
    }
    return t.tsArrayType(
      t.tsArrayType(
        getTypeOrRef(items[0])
      )
    );
  }

  // passing in {"items": [{"type":"string"}]}
  const detect = detectType(items.type);

  if (detect.type === 'array') {
    if (Array.isArray(items.items)) {
      return t.tsArrayType(
        t.tsArrayType(
          getTypeOrRef(items.items[0])
        )
      );
    } else {
      return t.tsArrayType(
        getArrayTypeFromItems(
          items.items
        )
      );
    }
  }


  return t.tsArrayType(
    getType(detect.type)
  );
}


export const detectType = (type: string | string[]) => {
  let optional = false;
  let theType = '';
  if (Array.isArray(type)) {
    if (type.length !== 2) {
      throw new Error('[getType(array length)] case not handled by transpiler. contact maintainers.')
    }
    const [nullableType, nullType] = type;
    if (nullType !== 'null') {
      throw new Error('[getType(null)] case not handled by transpiler. contact maintainers.')
    }
    theType = nullableType;
    optional = true;
  } else {
    theType = type;
  }

  return {
    type: theType,
    optional
  };
}

export const getTypeInfo = (info: JSONSchema) => {
  let type = undefined;
  let optional = undefined;

  if (Array.isArray(info.anyOf)) {
    // assuming 2nd is null, but let's check to ensure
    if (info.anyOf.length !== 2) {
      throw new Error('case not handled by transpiler. contact maintainers.')
    }
    const [nullableType, nullType] = info.anyOf;
    if (nullType?.type !== 'null') {
      throw new Error('[nullableType.type]: case not handled by transpiler. contact maintainers.')
    }
    if (!nullableType?.$ref) {
      if (nullableType.title) {
        type = t.tsTypeReference(t.identifier(nullableType.title));
      } else {
        throw new Error('[nullableType.title] case not handled by transpiler. contact maintainers.')
      }
    } else {
      type = getTypeFromRef(nullableType?.$ref);
    }
    optional = true;
  }


  if (typeof info.type === 'string') {
    if (info.type === 'array') {
      if (typeof info.items === 'object' && !Array.isArray(info.items)) {
        if (info.items.$ref) {
          type = getArrayTypeFromRef(info.items.$ref);
        } else if (info.items.title) {
          type = t.tsArrayType(
            t.tsTypeReference(
              t.identifier(info.items.title)
            )
          );
        } else if (info.items.type) {
          type = getArrayTypeFromItems(info.items);
        } else {
          throw new Error('[info.items] case not handled by transpiler. contact maintainers.')
        }
      } else {
        if (Array.isArray(info.items)) {
          type = getArrayTypeFromItems(info.items);
          // console.log(typeof info.items === 'object');
          // console.log(Array.isArray(info.items));
          // console.log(info);
        } else {
          throw new Error('[info.items] case not handled by transpiler. contact maintainers.')
        }
      }
    } else {
      const detect = detectType(info.type);
      type = getType(detect.type);
      optional = detect.optional;
    }
  }

  if (Array.isArray(info.type)) {
    // assuming 2nd is null, but let's check to ensure
    if (info.type.length !== 2) {
      throw new Error('please report this to maintainers (field type): ' + JSON.stringify(info, null, 2))
    }
    const [nullableType, nullType] = info.type;
    if (nullType !== 'null') {
      throw new Error('please report this to maintainers (field type): ' + JSON.stringify(info, null, 2))
    }

    if (nullableType === 'array' && typeof info.items === 'object' && !Array.isArray(info.items)) {

      if (info.items.type) {
        const detect = detectType(info.items.type);
        if (detect.type === 'array') {
          // wen recursion?
          type = t.tsArrayType(
            getArrayTypeFromItems(info.items)
          );
        } else {
          type = t.tsArrayType(
            getType(detect.type)
          );
        }
        optional = detect.optional;
      } else if (info.items.$ref) {
        type = getArrayTypeFromRef(info.items.$ref);
        // } else if (info.items.title) {
        //   type = t.tsArrayType(
        //     t.tsTypeReference(
        //       t.identifier(info.items.title)
        //     )
        //   );
      } else if (info.items.type) {
        type = getArrayTypeFromItems(info.items);
      } else {
        throw new Error('[info.items] case not handled by transpiler. contact maintainers.')
      }

    } else {
      const detect = detectType(nullableType);
      optional = detect.optional;
      if (detect.type === 'array') {
        type = getArrayTypeFromItems(
          info.items
        );
      } else {
        type = getType(detect.type);
      }

    }

    optional = true;
  }

  return {
    type,
    optional
  };

}


export const getType = (type: string) => {
  switch (type) {
    case 'string':
      return t.tsStringKeyword();
    case 'boolean':
      return t.tSBooleanKeyword();
    case 'integer':
      return t.tsNumberKeyword();
    default:
      throw new Error('contact maintainers [unknown type]: ' + type);
  }
}

export const getPropertyType = (
  context: RenderContext,
  schema: JSONSchema,
  prop: string
) => {
  const props = schema.properties ?? {};
  let info = props[prop];

  let type = null;
  let optional = !schema.required?.includes(prop);

  if (info.allOf && info.allOf.length === 1) {
    info = info.allOf[0];
  }

  if (typeof info.$ref === 'string') {
    type = getTypeFromRef(info.$ref)
  }

  const typeInfo = getTypeInfo(info);
  if (typeof typeInfo.optional !== 'undefined') {
    optional = typeInfo.optional;
  }
  if (typeof typeInfo.type !== 'undefined') {
    type = typeInfo.type;
  }

  if (!type) {
    throw new Error('cannot find type for ' + JSON.stringify(info))
  }

  if (schema.required?.includes(prop)) {
    optional = false;
  }

  return { type, optional };
};


export function getPropertySignatureFromProp(
  context: RenderContext,
  jsonschema: JSONSchema,
  prop: string,
  camelize: boolean
) {
  if (jsonschema.properties[prop].type === 'object') {
    if (jsonschema.properties[prop].title) {
      return propertySignature(
        camelize ? camel(prop) : prop,
        t.tsTypeAnnotation(
          t.tsTypeReference(t.identifier(jsonschema.properties[prop].title))
        )
      );
    } else {
      throw new Error('getPropertySignatureFromProp() contact maintainer');
    }
  }

  if (Array.isArray(jsonschema.properties[prop].allOf)) {
    const isOptional = !jsonschema.required?.includes(prop);
    const unionTypes = jsonschema.properties[prop].allOf.map(el => {
      if (el.title) return el.title;
      if (el.$ref) return getTypeStrFromRef(el.$ref);
      return el.type;
    });
    // @ts-ignore:next-line
    const uniqUnionTypes = [...new Set(unionTypes)];

    if (uniqUnionTypes.length === 1) {
      return propertySignature(
        camelize ? camel(prop) : prop,
        t.tsTypeAnnotation(
          t.tsTypeReference(
            t.identifier(uniqUnionTypes[0])
          )
        ),
        isOptional
      );
    } else {
      return propertySignature(
        camelize ? camel(prop) : prop,
        t.tsTypeAnnotation(
          t.tsUnionType(
            uniqUnionTypes.map(typ =>
              t.tsTypeReference(
                t.identifier(typ)
              )
            )
          )
        ),
        isOptional
      );
    }
  } else if (Array.isArray(jsonschema.properties[prop].oneOf)) {
    const isOptional = !jsonschema.required?.includes(prop);
    const unionTypes = jsonschema.properties[prop].oneOf.map(el => {
      if (el.title) return el.title;
      if (el.$ref) return getTypeStrFromRef(el.$ref);
      return el.type;
    });
    // @ts-ignore:next-line
    const uniqUnionTypes = [...new Set(unionTypes)];
    if (uniqUnionTypes.length === 1) {
      return propertySignature(
        camelize ? camel(prop) : prop,
        t.tsTypeAnnotation(
          t.tsTypeReference(
            t.identifier(uniqUnionTypes[0])
          )
        ),
        isOptional
      );
    } else {
      return propertySignature(
        camelize ? camel(prop) : prop,
        t.tsTypeAnnotation(
          t.tsUnionType(
            uniqUnionTypes.map(typ =>
              t.tsTypeReference(
                t.identifier(typ)
              )
            )
          )
        ),
        isOptional
      );
    }

  }

  try {
    getPropertyType(context, jsonschema, prop);
  } catch (e) {
    console.log(e);
    console.log(JSON.stringify(jsonschema, null, 2), prop);
  }

  const { type, optional } = getPropertyType(context, jsonschema, prop);
  return propertySignature(
    camelize ? camel(prop) : prop,
    t.tsTypeAnnotation(
      type
    ),
    optional
  );
}

export const getParamsTypeAnnotation = (
  context: RenderContext,
  jsonschema: any,
  camelize: boolean = true
): t.TSTypeAnnotation => {
  const keys = Object.keys(jsonschema.properties ?? {});

  if (!keys.length && jsonschema.$ref) {
    return t.tsTypeAnnotation(getTypeFromRef(jsonschema.$ref))
  }

  if (!keys.length) return undefined;

  const typedParams = keys.map(prop => getPropertySignatureFromProp(
    context,
    jsonschema,
    prop,
    camelize
  ));

  return t.tsTypeAnnotation(
    t.tsTypeLiteral(
      // @ts-ignore:next-line
      [
        ...typedParams
      ]
    )
  )
}

export const createTypedObjectParams = (
  context: RenderContext,
  jsonschema: JSONSchema,
  camelize: boolean = true
): (t.Identifier | t.Pattern | t.RestElement) => {

  const keys = Object.keys(jsonschema.properties ?? {});
  if (!keys.length) {
    // is there a ref?
    if (jsonschema.$ref) {
      const obj = context.refLookup(jsonschema.$ref);
      // If there is a oneOf, then we need to create a type for it
      if (obj?.oneOf) {
        // the actual type of the ref
        const refType = jsonschema.$ref.split('/').pop();
        const refName = camel(refType);
        const id = t.identifier(refName);
        id.typeAnnotation = t.tsTypeAnnotation(t.tsTypeReference(t.identifier(refType)));
        return id
      } else if (obj) {
        return createTypedObjectParams(
          context,
          obj,
          camelize
        );
      }
    }

    // no results...
    return;
  }

  const params = keys.map(prop => {
    return t.objectProperty(
      camelize ? t.identifier(camel(prop)) : t.identifier(prop),
      camelize ? t.identifier(camel(prop)) : t.identifier(prop),
      false,
      true
    );
  });

  const obj = t.objectPattern(
    [
      ...params
    ]
  );

  obj.typeAnnotation = getParamsTypeAnnotation(
    context,
    jsonschema,
    camelize
  );

  return obj;
};

/**
 * CamelCasedProperties<Extract<ExecuteMsg, { exec_on_module: unknown }>['exec_on_module']>
 */
export function createExtractTypeAnnotation(underscoreName: string, msgTitle: string) {
    return t.tsTypeAnnotation(
        t.tsTypeReference(
            t.identifier('CamelCasedProperties'),
            t.tsTypeParameterInstantiation([
                t.tsIndexedAccessType(
                    t.tsTypeReference(
                        t.identifier('Extract'),
                        t.tsTypeParameterInstantiation([
                            t.tsTypeReference(t.identifier(msgTitle)),
                            t.tsTypeLiteral([
                                t.tsPropertySignature(
                                    t.identifier(underscoreName),
                                    t.tsTypeAnnotation(t.tsUnknownKeyword())
                                )
                            ])
                        ])
                    ),
                    t.tsLiteralType(t.stringLiteral(underscoreName))
                )
            ])
        )
    );
}
