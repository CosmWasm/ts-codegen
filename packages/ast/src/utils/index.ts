export * from './babel';
export { OPTIONAL_FUNDS_PARAM } from './constants';
export { FIXED_EXECUTE_PARAMS } from './constants';
export { PROVIDER_TYPES } from './constants';
export * from './ref';
export * from './types';
