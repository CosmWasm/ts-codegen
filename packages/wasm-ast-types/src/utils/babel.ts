import * as t from '@babel/types';
import { snake } from "case";
import { Field, QueryMsg, ExecuteMsg } from '../types';
import { TSTypeAnnotation, TSExpressionWithTypeArguments, Identifier } from '@babel/types';
import { refLookup } from './ref';

// t.TSPropertySignature - kind?
export const propertySignature = (
    name: string,
    typeAnnotation: TSTypeAnnotation,
    optional: boolean = false
): t.TSPropertySignature => {
    return {
        type: 'TSPropertySignature',
        kind: "get",
        key: t.identifier(name),
        typeAnnotation,
        optional
    }
};

export const identifier = (name: string, typeAnnotation: t.TSTypeAnnotation, optional: boolean = false): t.Identifier => {
    const type = t.identifier(name);
    type.typeAnnotation = typeAnnotation;
    type.optional = optional;
    return type;
};

export const tsTypeOperator = (typeAnnotation: t.TSType, operator: string) => {
    const obj = t.tsTypeOperator(typeAnnotation);
    obj.operator = operator;
    return obj;
};

export const getMessageProperties = (msg) => {
    let results = [];
    let objs = [];
    if (msg.anyOf) { objs = msg.anyOf; }
    else if (msg.oneOf) { objs = msg.oneOf; }
    else if (msg.allOf) { objs = msg.allOf; }

    for (const obj of objs) {
      if(obj.properties){
        results.push(obj);
      } else{
        if(obj.$ref){
          const ref = refLookup(obj.$ref, msg)

          const refProps = getMessageProperties(ref);

          results = [...results, ...refProps];
        }
      }
    }

    return results;
}

export const tsPropertySignature = (
    key: t.Expression,
    typeAnnotation: t.TSTypeAnnotation,
    optional: boolean
) => {
    const obj = t.tsPropertySignature(key, typeAnnotation);
    obj.optional = optional;
    return obj
};

export const tsObjectPattern = (
    properties: (t.RestElement | t.ObjectProperty)[],
    typeAnnotation: t.TSTypeAnnotation
) => {
    const obj = t.objectPattern(
        properties
    );
    obj.typeAnnotation = typeAnnotation;
    return obj;
}

export const callExpression = (
    callee: t.Expression | t.V8IntrinsicIdentifier,
    _arguments: (t.Expression | t.SpreadElement | t.ArgumentPlaceholder)[],
    typeParameters: t.TSTypeParameterInstantiation
) => {
    const callExpr = t.callExpression(callee, _arguments);
    callExpr.typeParameters = typeParameters;
    return callExpr;
};

export const bindMethod = (name: string) => {
    return t.expressionStatement(
        t.assignmentExpression('=', t.memberExpression(
            t.thisExpression(),
            t.identifier(name)
        ),
            t.callExpression(
                t.memberExpression(
                    t.memberExpression(
                        t.thisExpression(),
                        t.identifier(name)
                    ),
                    t.identifier('bind')
                ),
                [
                    t.thisExpression()
                ]
            )
        )
    )
}

export const typedIdentifier = (name: string, typeAnnotation: TSTypeAnnotation, optional: boolean = false): Identifier => {
    const type = t.identifier(name);
    type.typeAnnotation = typeAnnotation;
    type.optional = optional;
    return type;
}

export const promiseTypeAnnotation = (name: string) => {
    return t.tsTypeAnnotation(
        t.tsTypeReference(
            t.identifier('Promise'),
            t.tsTypeParameterInstantiation(
                [
                    t.tsTypeReference(t.identifier(name))
                ]
            )
        )
    );
}

export const abstractClassDeclaration = (name: string, body: any[], implementsExressions: TSExpressionWithTypeArguments[] = [], superClass: t.Identifier = null) => {
  const declaration = classDeclaration(
    name,
    body,
    implementsExressions,
    superClass
  );
  declaration.abstract = true
  return declaration;
};

export const classDeclaration = (name: string, body: any[], implementsExressions: TSExpressionWithTypeArguments[] = [], superClass: t.Identifier = null) => {
    const declaration = t.classDeclaration(
        t.identifier(name),
        superClass,
        t.classBody(body)
    );
    if (implementsExressions.length) {
        declaration.implements = implementsExressions;
    }
    return declaration;
};


export const classProperty = (
    name: string,
    typeAnnotation: TSTypeAnnotation = null,
    isReadonly: boolean = false,
    isStatic: boolean = false,
    noImplicitOverride: boolean = false
) => {
    const prop = t.classProperty(t.identifier(name));
    if (isReadonly) prop.readonly = true;
    if (isStatic) prop.static = true;
    if (typeAnnotation) prop.typeAnnotation = typeAnnotation;
    if (noImplicitOverride) prop.override = true;
    return prop;
};

export const classPrivateProperty = (
  name: string,
  value: t.Expression = null,
  typeAnnotation: TSTypeAnnotation = null,
  isStatic: boolean = false
) => {
    const prop = t.classPrivateProperty( t.privateName(t.identifier(name)), value, [], isStatic);
    if (typeAnnotation) prop.typeAnnotation = typeAnnotation;
    return prop;
}

export const arrowFunctionExpression = (
    params: (t.Identifier | t.Pattern | t.RestElement)[],
    body: t.BlockStatement,
    returnType: t.TSTypeAnnotation,
    isAsync: boolean = false
) => {
    const func = t.arrowFunctionExpression(params, body, isAsync);
    if (returnType) func.returnType = returnType;
    return func;
};


export const recursiveNamespace = (names, moduleBlockBody) => {
    if (!names || !names.length) return moduleBlockBody;
    const name = names.pop();
    const body = [
        t.exportNamedDeclaration(
            t.tsModuleDeclaration(
                t.identifier(name),
                t.tsModuleBlock(recursiveNamespace(names, moduleBlockBody))
            )
        )
    ];
    return body;
};

export const arrayTypeNDimensions = (body, n) => {
    if (!n) return t.tsArrayType(body);
    return t.tsArrayType(
        arrayTypeNDimensions(body, n - 1)
    );
};

export const FieldTypeAsts = {
    string: () => {
        return t.tsStringKeyword()
    },
    array: (type) => {
        return t.tsArrayType(FieldTypeAsts[type]())
    },
    Duration: () => {
        return t.tsTypeReference(t.identifier('Duration'))
    },
    Height: () => {
        return t.tsTypeReference(t.identifier('Height'))
    },
    Coin: () => {
        return t.tsTypeReference(t.identifier('Coin'))
    },
    Long: () => {
        return t.tsTypeReference(t.identifier('Long'))
    }
};

export const shorthandProperty = (prop: string, typeAnnotation?: t.TSTypeAnnotation) => {
    return t.objectProperty(identifier(prop, typeAnnotation), t.identifier(prop), false, true);
};

export const objectPattern = (properties: t.ObjectProperty[], typeAnnotation?: t.TSTypeAnnotation) => {
    const obj = t.objectPattern(properties);
    if (typeAnnotation) {
      obj.typeAnnotation = typeAnnotation;
    }
    return obj;
}

/**
 * Retrieve the types from the object pattern properties.
 * Example:
 ```ts
  autoTypedObjectPattern([
    shorthandProperty(
      'moduleId',
      t.tsTypeAnnotation(t.tsStringKeyword())
    ),
    shorthandProperty(
      'accountId',
      t.tsTypeAnnotation(t.tsNumberKeyword())
    ),
  ])
 ```
 */
export const autoTypedObjectPattern = (properties: t.ObjectProperty[]) => {
  const propertyTypes = properties.map((prop) => {
    return propertySignature(
      // @ts-ignore
      prop.key.name,
      // @ts-ignore
      prop.key.typeAnnotation
    );
  })

  return objectPattern(
    properties,
    t.tSTypeAnnotation(t.tsTypeLiteral(propertyTypes))
  );
}

export const importStmt = (names: string[], path: string) => {
    return t.importDeclaration(
        names.map(name => t.importSpecifier(t.identifier(name), t.identifier(name))),
        t.stringLiteral(path));
};

export const importAs = (name: string, importAs: string, importPath: string) => {
    return t.importDeclaration(
        [
            t.importSpecifier(
                t.identifier(importAs),
                t.identifier(name)
            )
        ],
        t.stringLiteral(importPath)
    )
};

export const importAminoMsg = () => {
    return importStmt(['AminoMsg'], '@cosmjs/amino');
};

export const getFieldDimensionality = (field: Field) => {
    let typeName = field.type;
    const isArray = typeName.endsWith('[]');
    let dimensions = 0;
    if (isArray) {
        dimensions = typeName.match(/\[\]/g).length - 1;
        typeName = typeName.replace(/\[\]/g, '');
    }
    return {
        typeName,
        dimensions,
        isArray
    };
}

export const memberExpressionOrIdentifier = (names) => {
    if (names.length === 1) {
        return t.identifier(names[0])
    }
    if (names.length === 2) {
        const [b, a] = names;
        return t.memberExpression(
            t.identifier(a),
            t.identifier(b)
        );
    }
    const [name, ...rest] = names;

    return t.memberExpression(
        memberExpressionOrIdentifier(rest),
        t.identifier(name)
    )
};

export const memberExpressionOrIdentifierSnake = (names) => {
    if (names.length === 1) {
        return t.identifier(snake(names[0]))
    }
    if (names.length === 2) {
        const [b, a] = names;
        return t.memberExpression(
            t.identifier(snake(a)),
            t.identifier(snake(b))
        );
    }
    const [name, ...rest] = names;

    return t.memberExpression(
        memberExpressionOrIdentifierSnake(rest),
        t.identifier(snake(name))
    )
};

/**
 * If optional, return a conditional, otherwise just the expression
 */
export const optionalConditionalExpression = (test: t.Expression, expression: t.Expression, alternate: t.Expression, optional: boolean = false): t.Expression => {
    return optional
        ? t.conditionalExpression(
            test,
            expression,
            alternate
        )
        : expression
}

export const typeRefOrUnionWithUndefined = (identifier: t.Identifier, optional: boolean = false): t.TSType => {
    const typeReference = t.tsTypeReference(identifier)
    return optional
        ? t.tsUnionType([
            typeReference,
            t.tsUndefinedKeyword()
        ])
        : typeReference
}

export const parameterizedTypeReference = (identifier: string, from: t.TSType, omit: string | Array<string>) => {
    return t.tsTypeReference(
        t.identifier(identifier),
        t.tsTypeParameterInstantiation([
            from,
            typeof omit === 'string'
                ? t.tsLiteralType(t.stringLiteral(omit))
                : t.tsUnionType(omit.map(o => t.tsLiteralType(t.stringLiteral(o))))
        ])
    )
}

/**
 * omitTypeReference(t.tsTypeReference(t.identifier('Cw4UpdateMembersMutation'),),'args').....
 * Omit<Cw4UpdateMembersMutation, 'args'>
 */
export const omitTypeReference = (from: t.TSType, omit: string | Array<string>) => {
    return parameterizedTypeReference('Omit', from, omit)
}


export const pickTypeReference = (from: t.TSType, pick: string | Array<string>) => {
    return parameterizedTypeReference('Pick', from, pick)
}

export const RECORD_STRING_UNKNOWN_TYPE_ANNOTATION = t.tSTypeAnnotation(
  t.tsTypeReference(
    t.identifier('Record'),
    t.tsTypeParameterInstantiation([
      t.tsStringKeyword(),
      t.tsUnknownKeyword()
    ])
  )
)
