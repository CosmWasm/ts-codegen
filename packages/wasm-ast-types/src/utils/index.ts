export * from './babel';
export * from './types';
export * from './ref';
export { OPTIONAL_FUNDS_PARAM } from './constants';
export { OPTIONAL_FIXED_EXECUTE_PARAMS } from './constants';
