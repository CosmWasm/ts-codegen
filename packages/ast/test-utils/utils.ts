import generate from '@babel/generator';
import { RenderOptions } from '@cosmwasm/ts-codegen-ast/types';
import { ExecuteMsg, IDLObject, JSONSchema, QueryMsg } from '@cosmwasm/ts-codegen-types';
import { readFileSync } from 'fs';
import { sync as glob } from 'glob';
import { join } from 'path';

import { RenderContext } from '../src/context';

export const expectCode = (ast: any): void => {
  expect(
    generate(ast).code
  ).toMatchSnapshot();
};

export const printCode = (ast: any): void => {
  console.log(
    generate(ast).code
  );
};

export const makeContext = (
  schema: JSONSchema,
  options?: RenderOptions,
  responses?: Record<string, JSONSchema>
) => {
  return new RenderContext({
    schemas: [schema],
    responses
  }, options);
};

interface GlobContractLegacy {
  name: `/${string}.json`;
  content: JSONSchema;
}

interface GlobContract {
  name: `/${string}.json`;
  content: IDLObject;
}

const fixtureDir = join(__dirname, '../../../__fixtures__');
const globCache: Record<string, GlobContractLegacy[] | GlobContract[]> = {};

export const globIdlBasedContracts = (p: string): GlobContract[] => {
  if (globCache[p]) return globCache[p] as GlobContract[];
  // @ts-ignore
  const contracts: GlobContract[] = glob(join(fixtureDir, p, '/*.json')).map(file => {
    return {
      name: file.split(join('__fixtures__', p))[1],
      content: JSON.parse(readFileSync(file, 'utf-8'))
    };
  });
  globCache[p] = contracts;
  return contracts;
};

export const globLegacyContracts = (p: string): GlobContractLegacy[] => {
  if (globCache[p]) return globCache[p] as GlobContractLegacy[];
  // @ts-ignore
  const contracts: GlobContractLegacy[] = glob(join(fixtureDir, p, '/*.json')).map(file => {
    return {
      name: file.split(join('__fixtures__', p))[1],
      content: JSON.parse(readFileSync(file, 'utf-8'))
    };
  });
  globCache[p] = contracts;
  return contracts;
};

export const getMsgExecuteLegacyFixture = (
  p: string,
  name: `/${string}.json`
): ExecuteMsg => {
  const contracts = globLegacyContracts(p);
  const contract = contracts.find(a => a.name === name);
  return contract.content as ExecuteMsg;
};

export const getMsgQueryLegacyFixture = (
  p: string,
  name: `/${string}.json`
): QueryMsg => {
  const contracts = globLegacyContracts(p);
  const contract = contracts.find(a => a.name === name);
  return contract.content as QueryMsg;
};

export const getLegacyFixture = (
  p: string,
  name: `/${string}.json`
): JSONSchema => {
  const contracts = globLegacyContracts(p);
  const contract = contracts.find(a => a.name === name);
  return contract.content as JSONSchema;
};
