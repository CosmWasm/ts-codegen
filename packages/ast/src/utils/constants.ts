import { identifier } from './babel';
import * as t from '@babel/types';

export const OPTIONAL_FUNDS_PARAM = identifier(
  '_funds',
  t.tsTypeAnnotation(t.tsArrayType(t.tsTypeReference(t.identifier('Coin')))),
  true
);
export const OPTIONAL_FEE_PARAM = identifier(
  'fee',
  t.tsTypeAnnotation(
    t.tsUnionType([
      t.tsNumberKeyword(),
      t.tsTypeReference(t.identifier('StdFee')),
      t.tsLiteralType(t.stringLiteral('auto'))
    ])
  ),
  true
);
export const OPTIONAL_MEMO_PARAM = identifier(
  'memo',
  t.tsTypeAnnotation(t.tsStringKeyword()),
  true
);

export const FIXED_EXECUTE_PARAMS = [
  OPTIONAL_FEE_PARAM,
  OPTIONAL_MEMO_PARAM,
  OPTIONAL_FUNDS_PARAM
];

export const PROVIDER_TYPES = {
  SIGNING_CLIENT_TYPE: "client",
  QUERY_CLIENT_TYPE: "queryClient",
  MESSAGE_COMPOSER_TYPE : 'message-composer',
  PROVIDER_TYPE : 'provider'
};
