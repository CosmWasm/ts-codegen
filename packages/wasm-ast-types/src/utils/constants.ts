import { identifier } from './babel';
import * as t from '@babel/types';

export const OPTIONAL_FUNDS_PARAM = identifier(
  '_funds',
  t.tsTypeAnnotation(t.tsArrayType(t.tsTypeReference(t.identifier('Coin')))),
  true
);

export const FEE_PARAM = identifier(
  'fee',
  t.tsTypeAnnotation(
    t.tsUnionType([
      t.tsNumberKeyword(),
      t.tsTypeReference(t.identifier('StdFee')),
      t.tsLiteralType(t.stringLiteral('auto'))
    ])
  )
);

export const OPTIONAL_FEE_PARAM: t.Identifier = {
  ...FEE_PARAM,
  optional: true
};

export const OPTIONAL_MEMO_PARAM = identifier(
  'memo',
  t.tsTypeAnnotation(t.tsStringKeyword()),
  true
);

export const OPTIONAL_FIXED_EXECUTE_PARAMS = [
  OPTIONAL_FEE_PARAM,
  OPTIONAL_MEMO_PARAM,
  OPTIONAL_FUNDS_PARAM
];
