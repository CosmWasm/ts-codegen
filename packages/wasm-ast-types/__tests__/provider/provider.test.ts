import {
  createGettingProviders,
  createIContractsContext,
  createProvider
} from '../../src/provider/provider';
import { PROVIDER_TYPES } from '../../src/utils/constants';
import { expectCode } from '../../test-utils';

it('execute class', () => {
  const info = {};

  info[PROVIDER_TYPES.SIGNING_CLIENT_TYPE] = {
    classname: 'WhitelistClient'
  };

  info[PROVIDER_TYPES.QUERY_CLIENT_TYPE] = {
    classname: 'WhitelistQueryClient'
  };

  info[PROVIDER_TYPES.MESSAGE_COMPOSER_TYPE] = {
    classname: 'WhitelistMessageComposer'
  };

  expectCode(createProvider('Whitelist', info));
});

it('execute class without message composer', () => {
  const info = {};

  info[PROVIDER_TYPES.SIGNING_CLIENT_TYPE] = {
    classname: 'WhitelistClient'
  };

  info[PROVIDER_TYPES.QUERY_CLIENT_TYPE] = {
    classname: 'WhitelistQueryClient'
  };

  expectCode(createProvider('Whitelist', info));
});

it('create IContractsContext', () => {
  const info = {
    Whitelist: {},
    Marketplace: {}
  };

  info['Whitelist'][PROVIDER_TYPES.SIGNING_CLIENT_TYPE] = {
    classname: 'WhitelistClient'
  };

  info['Whitelist'][PROVIDER_TYPES.QUERY_CLIENT_TYPE] = {
    classname: 'WhitelistQueryClient'
  };

  info['Marketplace'][PROVIDER_TYPES.SIGNING_CLIENT_TYPE] = {
    classname: 'MarketplaceClient'
  };

  expectCode(createIContractsContext(info));
});

it('create getProviders', () => {
  const info = {
    Whitelist: {},
    Marketplace: {}
  };

  info['Whitelist'][PROVIDER_TYPES.SIGNING_CLIENT_TYPE] = {
    classname: 'WhitelistClient'
  };

  info['Whitelist'][PROVIDER_TYPES.QUERY_CLIENT_TYPE] = {
    classname: 'WhitelistQueryClient'
  };

  info['Marketplace'][PROVIDER_TYPES.SIGNING_CLIENT_TYPE] = {
    classname: 'MarketplaceClient'
  };

  expectCode(createGettingProviders(info));
});
