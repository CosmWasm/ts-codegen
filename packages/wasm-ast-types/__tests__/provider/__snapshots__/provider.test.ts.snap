// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create IContractsContext 1`] = `
"export interface IContractsContext {
  whitelist: ISigningClientProvider<WhitelistClient> & IQueryClientProvider<WhitelistQueryClient>;
  marketplace: ISigningClientProvider<MarketplaceClient>;
}"
`;

exports[`create getProviders 1`] = `
"export const getProviders = (address?: string, cosmWasmClient?: CosmWasmClient, signingCosmWasmClient?: SigningCosmWasmClient) => ({
  whitelist: new Whitelist({
    address,
    cosmWasmClient,
    signingCosmWasmClient
  }),
  marketplace: new Marketplace({
    address,
    cosmWasmClient,
    signingCosmWasmClient
  })
});"
`;

exports[`execute class 1`] = `
"export class Whitelist extends ContractBase<WhitelistClient, WhitelistQueryClient, WhitelistMessageComposer> {
  constructor({
    address,
    cosmWasmClient,
    signingCosmWasmClient
  }: IContractConstructor) {
    super(address, cosmWasmClient, signingCosmWasmClient, WhitelistClient, WhitelistQueryClient, WhitelistMessageComposer);
  }

}"
`;

exports[`execute class without message composer 1`] = `
"export class Whitelist extends ContractBase<WhitelistClient, WhitelistQueryClient, IEmptyClient> {
  constructor({
    address,
    cosmWasmClient,
    signingCosmWasmClient
  }: IContractConstructor) {
    super(address, cosmWasmClient, signingCosmWasmClient, WhitelistClient, WhitelistQueryClient, undefined);
  }

}"
`;
